### PR TITLE
[CommandPropertyToAttribute] #254

### PR DESCRIPTION
**Pull request by JohJohan**

### CODE REVIEW

**Comment by johanadivare** [comment]
The test are failing on `Rector\Core\Exception\ShouldNotHappenException: Chain method calls cannot be removed this way. It would remove the whole tree of calls. Remove them manually by creating new parent node with no following method.` i dont know how to remove the node for hidden setter

**Comment by JohJohan** [comment]
@samsonasik do you know why https://github.com/rectorphp/rector-symfony/pull/259#issuecomment-1289051683?

**Comment by samsonasik** [comment]
@JohJohan you were trying to remove partial method call from fluent call. You need to create new parent node or replace root node with new Node instead.

**Comment by JohJohan** [comment]
> @JohJohan you were trying to remove partial method call from fluent call. You need to create new parent node or replace root node with new Node instead.

Okay, i looked into it but i'm affraid i'm not able to fix it. I tried looking at our docs or other code for examples. I think we could use `createLocalMethodCall` but that needs a new method name

**Comment by johanadivare** [comment]
> > @JohJohan you were trying to remove partial method call from fluent call. You need to create new parent node or replace root node with new Node instead.
> 
> Okay, i looked into it but i'm affraid i'm not able to fix it. I tried looking at our docs or other code for examples. I think we could use `createLocalMethodCall` but that needs a new method name

@samsonasik maybe you could give another hint on where to find the solution

**Comment by samsonasik** [comment]
@JohJohan you need to replace parent node var with the node, I made a commit for it :) https://github.com/rectorphp/rector-symfony/pull/259/commits/d4d075e18dba8dac10f59c9437f6a35fb51bcfdd

**Comment by johanadivare** [comment]
@samsonasik thanks i will fix the phpstan issues

**Comment by johanadivare** [comment]
@samsonasik i cleanup the commit history by rebasing and create 1 commit for my changes

**Review by samsonasik — APPROVED** [review]

File: src/Rector/Return_/CommandConstantReturnCodeRector.php [hunk 1]
```diff
@@ -82,17 +81,16 @@ public function refactor(Node $node): ?Node
             return null;
         }
 
-        if (! $this->nodeNameResolver->isName($node, 'execute')) {
+        if (! $classReflection->hasMethod('execute')) {
```
Comment (samsonasik) [inline src/Rector/Return_/CommandConstantReturnCodeRector.php:84]: This is wrong, hasMethod verify method exists, not check existing method that return parent of it

File: src/Rector/Return_/CommandConstantReturnCodeRector.php [hunk 2]
```diff
@@ -65,11 +64,11 @@ protected function execute(InputInterface $input, OutputInterface $output): int
      */
     public function getNodeTypes(): array
     {
-        return [ClassMethod::class];
+        return [Return_::class];
```
Comment (samsonasik) [inline src/Rector/Return_/CommandConstantReturnCodeRector.php:67]: The old ClassMethod is the correct way, as directly check ClassMethod name

**Review by samsonasik — CHANGES_REQUESTED** [review]
Please revert `CommandConstantReturnCodeRector` to use ClassMethod node, as before, see prev comment

File: src/Rector/Return_/CommandConstantReturnCodeRector.php (hunk 2 above)
Comment (johanadivare) [reply src/Rector/Return_/CommandConstantReturnCodeRector.php]: I think i resolved the merge incorrectly at some point thanks for finding it.

**Comment by johanadivare** [comment]
@samsonasik i revert my invalid changes and squashed my commits again

File: src/Rector/Class_/CommandPropertyToAttributeRector.php [hunk 3]
```diff
@@ -118,34 +203,74 @@ public function refactor(Node $node): ?Node
...
+
+            if (! $this->isName($node->name, 'setHidden')) {
+                return null;
+            }
+
+            if (! $node->getArgs()[0]->value instanceof ConstFetch) {
+                return null;
+            }
+
+            $commandHidden = $node->getArgs()[0]
```
Comment (samsonasik) [inline src/Rector/Class_/CommandPropertyToAttributeRector.php:263]: multiple `getArgs()` can be saved to variable, also, if you have `$node->isFirstCallable()` check, you can just use `$node->args`, the `@var Arg $firstArg` is fine after that

Comment (johanadivare) [reply src/Rector/Class_/CommandPropertyToAttributeRector.php]: I defined a variable for getArgs, i had a look at `$node->isFirstCallable()` but its not a function on `$node`?

Comment (johanadivare) [reply src/Rector/Class_/CommandPropertyToAttributeRector.php]: Ah its probably `isFirstClassCallable`

Comment (johanadivare) [reply src/Rector/Class_/CommandPropertyToAttributeRector.php]: Was discussed in https://github.com/rectorphp/rector-symfony/pull/279/files#r1029199016

Comment (johanadivare) [reply src/Rector/Class_/CommandPropertyToAttributeRector.php]: Its ready now

**Review by samsonasik — APPROVED** [review]

**Review by samsonasik — CHANGES_REQUESTED** [review]
One more check on Arg value exist and its value

File: tests/Rector/Class_/CommandPropertyToAttributeRector/Fixture/commandWithDescriptionAndAliasesAndHiddenAndHelp.php.inc [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+    public static $defaultName = 'sunshine';
+
+    /** @var string|null */
+    public static $defaultDescription = 'Let the sun shine!';
+
+    public function configure(): void
+    {
+        $this
+            ->setAliases(['sun', 'shine'])
+            ->setHidden(false)
```
[the remark this answers is not in the document]
Comment (johanadivare) [reply tests/Rector/Class_/CommandPropertyToAttributeRector/Fixture/commandWithDescriptionAndAliasesAndHiddenAndHelp.php.inc:17]: You mean what about `setHidden()` which would set the hidden to `true` right now the code fails so i will add some checks

[the remark this answers is not in the document]
Comment (johanadivare) [reply tests/Rector/Class_/CommandPropertyToAttributeRector/Fixture/commandWithDescriptionAndAliasesAndHiddenAndHelp.php.inc:17]: I pushed some code let me know what you think. Also good catch!

[the remark this answers is not in the document]
Comment (samsonasik) [reply tests/Rector/Class_/CommandPropertyToAttributeRector/Fixture/commandWithDescriptionAndAliasesAndHiddenAndHelp.php.inc:17]: On 2 conditions:

- no arg (use default value)
- true arg

[the remark this answers is not in the document]
Comment (johanadivare) [reply tests/Rector/Class_/CommandPropertyToAttributeRector/Fixture/commandWithDescriptionAndAliasesAndHiddenAndHelp.php.inc:17]: I fixed phpstan issue

File: tests/Rector/Class_/CommandPropertyToAttributeRector/Fixture/commandWithHidden.php.inc [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+
+class CommandWithHidden extends \Symfony\Component\Console\Command\Command
+{
+    /** @var string|null */
+    public static $defaultName = 'sunshine';
+
+    public function configure(): void
+    {
+        $this
+            ->setHidden();
```
Comment (samsonasik) [inline tests/Rector/Class_/CommandPropertyToAttributeRector/Fixture/commandWithHidden.php.inc:13]: setHidden(true) test case need to be added, are they need to be removed whatever value pass on it?

Comment (johanadivare) [reply tests/Rector/Class_/CommandPropertyToAttributeRector/Fixture/commandWithHidden.php.inc:13]: Added testcase with `setHidden(true)` false case is handled in https://github.com/rectorphp/rector-symfony/pull/259/files#diff-1ce73e8d15ebe88b93e9a7d208697815fcadb4ab67e8067a0e646e9bee4141db

What do you mean with `are they need to be removed whatever value pass on it?`?

Comment (samsonasik) [reply tests/Rector/Class_/CommandPropertyToAttributeRector/Fixture/commandWithHidden.php.inc:13]: There are 3 possible conditions:

- setHidden()
- setHidden(false)
- setHidden(true)

Are they need to be removed so need check against Arg value ?

Comment (johanadivare) [reply tests/Rector/Class_/CommandPropertyToAttributeRector/Fixture/commandWithHidden.php.inc:13]: I think what you mean is do we need to remove `setHidden` function on all these cases and move them to the `AsCommand` attribute? I would say yes. But not sure that is what you mean

Comment (samsonasik) [reply tests/Rector/Class_/CommandPropertyToAttributeRector/Fixture/commandWithHidden.php.inc:13]: Yes, that's what I mean 👍 , it seems test added now, looks good 👍

**Review by samsonasik — APPROVED** [review]

**Review by samsonasik — CHANGES_REQUESTED** [review]
Just general improvement:

1. are the `setHidden()` and `setAliases()` only should be called in `configure()` method, if yes, you may lookup only on `configure()` ClassMethod to improve performance.

```php
$classMethod = $class->getMethod('configure');

if (! $classMethod instanceof \PhpParser\Node\Stmt\ClassMethod) {
    return;
}

// traverse $classMethod->stmts only here...
```

3. After Node changed inside `traverseNodesWithCallable()`, `return \PhpParser\NodeTraverser\NodeTraverser::STOP_TRAVERSAL;` should be called to avoid unnecessary another finding.

You can check its usage in https://github.com/rectorphp/rector-src/blob/a9372bb3e90aa966412b2deab075a44769bb119c/src/NodeAnalyzer/ParamAnalyzer.php

3. If you only need to locate the MethodCall in Expression, you just need to loop the $clasMethod->stmts :

```php
foreach ((array) $classMethod->stmts as $stmt) {
    if (! $stmt instanceof \PhpParser\Node\Stmt\Expression) {
         // stop, possibly there is if else
         return;
    }
}

/** @var \PhpParser\Node\Stmt\Expression $stmts[] */
$stmts =(array) $classMethod->stmts; 
foreach ($stmts as $stmt) {
    if (! $stmt->expr instanceof MethodCall) {
          continue;
     }

     // check setAliases and setHidden() here
}
```

**Comment by johanadivare** [comment]
> 

I did 1/2 but for 3 how do i get the value from the Expression and check that it is SetHidden for example?

**Comment by samsonasik** [comment]
If it always in Expression, you can verify that Expression->expr is MethodCall, then verify after that, but I think that can be next iteration improvement. 

Current implementation looks good 👍

**Review by samsonasik — APPROVED** [review]

File: src/Rector/Class_/CommandPropertyToAttributeRector.php [hunk 6]
```diff
@@ -118,34 +206,104 @@ public function refactor(Node $node): ?Node
...
-    private function createAttributeGroup(string $defaultName, ?string $defaultDescription): AttributeGroup
+    private function resolveAliases(Class_ $class): ?Array_
     {
-        $attributeGroup = $this->phpAttributeGroupFactory->createFromClass(self::ATTRIBUTE);
+        $commandAliases = null;
+        $classMethod = $class->getMethod('configure');
+        if (! $classMethod instanceof ClassMethod) {
+            return null;
+        }
+        $this->traverseNodesWithCallable($class->stmts, function (Node $node) use (&$commandAliases) {
```
Comment (samsonasik) [inline src/Rector/Class_/CommandPropertyToAttributeRector.php:219]: Oh, it need to be `$classMethod->stmts`:

```suggestion
        $this->traverseNodesWithCallable((array) $classMethod->stmts, function (Node $node) use (&$commandAliases) {
```

File: src/Rector/Class_/CommandPropertyToAttributeRector.php [hunk 7]
```diff
@@ -118,34 +206,104 @@ public function refactor(Node $node): ?Node
...
-        if (count($property->props) !== 1) {
+        $commandHidden = null;
+        $classMethod = $class->getMethod('configure');
+        if (! $classMethod instanceof ClassMethod) {
             return null;
         }
-        $propertyProperty = $property->props[0];
-        if ($propertyProperty->default instanceof String_) {
-            return $propertyProperty->default->value;
+        $this->traverseNodesWithCallable($class->stmts, function (Node $node) use (&$commandHidden) {
```
Comment (samsonasik) [inline src/Rector/Class_/CommandPropertyToAttributeRector.php:264]: Oh, it need to be `$classMethod->stmts`:

```suggestion
        $this->traverseNodesWithCallable((array) $classMethod->stmts, function (Node $node) use (&$commandHidden) {
```

File: src/Rector/Class_/CommandDescriptionToPropertyRector.php [hunk 8]
```diff
@@ -108,33 +111,42 @@ private function resolveCommandDescription(Class_ $class): ?Node
     private function resolveCommandDescriptionFromSetDescription(Class_ $class): ?Node
     {
         $commandDescription = null;
-
+        $classMethod = $class->getMethod('configure');
+        if (! $classMethod instanceof ClassMethod) {
+            return null;
+        }
         $this->traverseNodesWithCallable($class->stmts, function (Node $node) use (&$commandDescription) {
```
Comment (samsonasik) [inline src/Rector/Class_/CommandDescriptionToPropertyRector.php:118]: Oh, it need to be `$classMethod->stmts`:

```suggestion
        $this->traverseNodesWithCallable((array) $classMethod->stmts, function (Node $node) use (&$commandDescription) {
```

**Review by samsonasik — CHANGES_REQUESTED** [review]
Tge traverse to use `(array) $classMethod->stmts` instead of `$class->stmts`

**Comment by samsonasik** [comment]
You can verify it only on Expression with verify:

```php
$expressionStmts = array_filter((array) $classMethod->stmts, fn (\PhpParser\Node\Stmt $stmt): bool => $stmt instanceof \PhpParser\Node\Stmt\Expression);
if (count((array) $classMethod->stmts) > count($expressionStmts)) {

    return;

}
```

so the traverse will only on Expression

File: src/Rector/Class_/CommandPropertyToAttributeRector.php (hunk 6 above)
Comment (johanadivare) [reply src/Rector/Class_/CommandPropertyToAttributeRector.php:219]: ++

File: src/Rector/Class_/CommandPropertyToAttributeRector.php (hunk 7 above)
Comment (johanadivare) [reply src/Rector/Class_/CommandPropertyToAttributeRector.php:264]: ++

File: src/Rector/Class_/CommandDescriptionToPropertyRector.php (hunk 8 above)
Comment (johanadivare) [reply src/Rector/Class_/CommandDescriptionToPropertyRector.php:118]: ++

**Comment by johanadivare** [comment]
> ```
> 
>     if (! $stmt instanceof \PhpParser\Node\Stmt\Expression) {
>          // stop, possibly there is if else
>          return;
>     }
> ```

I dont see the benefit yet for checking if it is a `Expression` but it would look something like:
```
 foreach ((array) $classMethod->stmts as $stmt) {
            if (!$stmt instanceof \PhpParser\Node\Stmt\Expression) {
                // stop, possibly there is if else
                continue;
            }

            if (!$stmt->expr instanceof MethodCall) {
                continue;
            }


            if (!$this->isName($stmt->expr->name, 'setAliases')) {
                continue;
            }

            /** @var Arg $arg */
            $arg = $stmt->expr->args[0];
            if (!$arg->value instanceof Array_) {
                return null;
            }

            $commandAliases = $arg->value;

            $parentNode = $stmt->expr->getAttribute(AttributeKey::PARENT_NODE);
            if ($parentNode instanceof MethodCall) {
                $parentNode->var = $stmt->expr->var;
            } else {
                $this->removeNode($stmt->expr);
            }
        }
```
Right?

Or do you still want me to use `traverseNodesWithCallable` but with the check you mentioned so i will only traverse if it is a Expression?

**Comment by samsonasik** [comment]
If you want to loop, you need to double loop, once is for verify if the stmts only Expression with MethodCall Expr, next is to apply changes, directly change will make it already changed while stop in next If statement, which invalid.

Using array_filter() with compare count can be both safer and less code I think.

**Comment by johanadivare** [comment]
> If you want to loop, you need to double loop, once is for verify if the stmts only Expression with MethodCall Expr, next is to apply changes, directly change will make it already changed while stop in next If statement, which invalid.
> 
> Using array_filter() with compare count can be both safer and less code I think, traverse is find, but verify first as it apply change during traversion.

I be honest and dont know 100% sure what is going on but i tried:
```php

        $expressionStmts = array_filter(
            (array)$classMethod->stmts,
            fn(\PhpParser\Node\Stmt $stmt): bool => $stmt instanceof \PhpParser\Node\Stmt\Expression
        );
        if (count((array)$classMethod->stmts) > count($expressionStmts)) {
            return null;
        }
        foreach ((array) $classMethod->stmts as $stmt) {
            if (!$stmt->expr instanceof MethodCall) {
                continue;
            }
            if (!$this->isName($stmt->expr->name, 'setAliases')) {
                continue;
            }

            /** @var Arg $arg */
            $arg = $stmt->expr->args[0];
            if (!$arg->value instanceof Array_) {
                return null;
            }

            $commandAliases = $arg->value;

            $parentNode = $stmt->expr->getAttribute(AttributeKey::PARENT_NODE);
            if ($parentNode instanceof MethodCall) {
                $parentNode->var = $stmt->expr->var;
            } else {
                $this->removeNode($stmt->expr);
            }
        }
```

But this didn't work see test output:
```diff
 
-#[\Symfony\Component\Console\Attribute\AsCommand('sunshine', 'Let the sun shine!', ['sun', 'shine'], false)]
+#[\Symfony\Component\Console\Attribute\AsCommand('sunshine', 'Let the sun shine!', [], false)]
 class CommandWithDescriptionAndAliasesAndHiddenAndHelp extends \Symfony\Component\Console\Command\Command
 {
     public function configure(): void
     {
         $this
+            ->setAliases(['sun', 'shine'])
             ->setHelp('Let the sun shine! Its that simple')
         ;
     }

```

**Review by samsonasik — APPROVED** [review]

**Comment by samsonasik** [comment]
Thank you @JohJohan

**Pull request merged**